petab_select/candidate_space.py

```diff
@@ -529,6 +602,628 @@ def iterate():
+        """Update self.method to the method."""
+        if method not in [Method.FORWARD, Method.LATERAL, Method.BACKWARD]:
+            raise NotImplementedError(
+                f'FAMoS direction must be either `Method.FORWARD`, `Method.BACKWARD` or `Method.LATERAL`, not {method}. \
+                Check if the method_scheme scheme provided is correct.'
+            )
+
+        self.method = method
+
+    def switch_inner_candidate_space(self, history):
+        """Switch self.inner_candidate_space to the candidate space of
+        the current self.method."""
+
+        self.inner_candidate_space = self.inner_candidate_spaces[self.method]
+        # reset the next inner candidate space with the current history of all checked models
+        self.inner_candidate_space.reset(
+            predecessor_model=self.predecessor_model, exclusions=history.keys
+        )
+        # FIXME Check that problem.calibrated_models is excluded.
+
+    def jump_to_most_distant(self):
+        """Jump to most distant model with respect to the history of all
+        calibrated models."""
+
+        new_init_model = self.get_most_distant()
+
+        logging.info("JUMPING: ", new_init_model.parameters)
+        # breakpoint()
+
+        # if model not appropriate make it so by adding the first
+        # critical parameter from each critical parameter set
+        if not self.check_critical(new_init_model):
+            for critical_set in self.critical_parameter_sets:
+                new_init_model.parameters[critical_set[0]] = ESTIMATE
+
+        self.update_method(self.initial_method)
+
+        self.number_of_reattempts -= 1
+        self.jumped_to_most_distant = True
+
+        self.predecessor_model = new_init_model
+        self.best_model_of_current_run = new_init_model
+        if self.summary_tsv is not None:
+            with open(self.summary_tsv, 'a') as f:
+                writer = csv.writer(f, delimiter='\t')
+                writer.writerow(
+                    [
+                        "Jumped to the most distant model.",
+                        *([''] * 5),
+                    ]
+                )
+
+    # TODO Fix for non-famos model subspaces. FAMOS easy beacuse of only 0;ESTIMATE
+    def get_most_distant(self) -> Model:
+        """
+        Get most distant model to all the checked models. We take models from the
+        sorted list of best models (self.best_models) and construct complements of
+        these models. For all these complements we compute the distance in number of
+        different estimated parameters to all models from history. For each complement
+        we take the minimum of these distances as it's distance to history. Then we
+        choose the complement model with the maximal distance to history.
+        TODO:
+        Next we check if this model is contained in any subspace. If so we choose it.
+        If not we choose the model in a subspace that has least distance to this
+        complement model.
```

Comment:
check

Reply:
One of the bigger TODOs to generalize get_most_distant

Reply:
Could you label all `TODO`s that are related to generalizing to multiple subspaces, e.g. `TODO (generalize FAMoS to multiple subspaces)`. An error or warning is currently emitted if multiple subspaces are used, right?